src/identity/identity-wallet.ts

```diff
@@ -1122,4 +1288,136 @@ export class IdentityWallet implements IIdentityWallet {
       isStateGenesis: false
     });
   }
+
+  /** {@inheritdoc IIdentityWallet.transitState} */
+  async transitState(
+    did: DID,
+    oldTreeState: TreeState,
+    isOldStateGenesis: boolean,
+    ethSigner: Signer
+  ): Promise<string> {
+    const newTreeModel = await this.getDIDTreeModel(did);
+    const claimsRoot = await newTreeModel.claimsTree.root();
+    const rootOfRoots = await newTreeModel.rootsTree.root();
+    const revocationRoot = await newTreeModel.revocationTree.root();
+
+    const newTreeState: TreeState = {
+      revocationRoot,
+      claimsRoot,
+      state: newTreeModel.state,
+      rootOfRoots
+    };
+
+    const userId = DID.idFromDID(did);
+
+    let proof;
+    const isEthIdentity = isEthereumIdentity(did); // don't generate proof for ethereum identities
+
+    if (!isEthIdentity && this._prover) {
+      // generate the proof
+      const authInfo = await this._inputsGenerator.prepareAuthBJJCredential(did, oldTreeState);
+      const challenge = Poseidon.hash([oldTreeState.state.bigInt(), newTreeState.state.bigInt()]);
+
+      const signature = await this.signChallenge(challenge, authInfo.credential);
+
+      const circuitInputs = new StateTransitionInputs();
+      circuitInputs.id = userId;
+
+      circuitInputs.signature = signature;
+      circuitInputs.isOldStateGenesis = isOldStateGenesis;
+
+      const authClaimIncProofNewState = await this.generateCredentialMtp(
+        did,
+        authInfo.credential,
+        newTreeState
+      );
+
+      circuitInputs.newTreeState = authClaimIncProofNewState.treeState;
+      circuitInputs.authClaimNewStateIncProof = authClaimIncProofNewState.proof;
+
+      circuitInputs.oldTreeState = oldTreeState;
+      circuitInputs.authClaim = {
+        claim: authInfo.coreClaim,
+        incProof: authInfo.incProof,
+        nonRevProof: authInfo.nonRevProof
+      };
+
+      const inputs = circuitInputs.inputsMarshal();
+
+      proof = await this._prover.generate(inputs, CircuitId.StateTransition);
+    }
+
+    const oldUserState = oldTreeState.state;
+    const newUserState = newTreeState.state;
+    const userStateTransitionInfo: UserStateTransitionInfo = {
+      userId,
+      oldUserState,
+      newUserState,
+      isOldStateGenesis,
+      methodId: BigInt(1),
+      methodParams: '0x'
+    } as UserStateTransitionInfo;
+
+    let txId;
+    if (!isEthIdentity) {
+      txId = await this._storage.states.publishState(proof, ethSigner);
+    } else {
+      txId = await this._storage.states.publishStateGeneric(ethSigner, userStateTransitionInfo);
+    }
+    await this.updateIdentityState(did, true, newTreeState);
+
+    return txId;
+  }
+
+  /** {@inheritdoc IIdentityWallet.addBJJAuthjjCredential} */
+  async addBJJAuthjjCredential(
```

Comment:
Is this function specific for Eth identity? 
If yes maybe better to call it `addBJJAuthjjCredentialToEthIdentity` ? 
And check `isEthereumIdentity(did)` ?;

Reply:
This function will be used also for key rotation for identities in general